### DISCUSSION TITLE
UI redesign

### DISCUSSION THREAD

**Discussion by hipersayanX** [webcamoid/webcamoid#517] [Ideas]
Here I'll comment about of all upcoming changes in the UI of Webcamoid.  
In the pasts weeks, I been redesigning the UI to give it a more fresh and consistent look. In the next video you can see the difference between the old style and the new one:  

https://streamable.com/8klwm

The first problem with the old look is that I've used the [Qt's Universal style](https://doc.qt.io/qt-5/qtquickcontrols2-universal.html) by default, the reason to use this style was due to serious inconsistencies between platforms using the default Qt theme. Here you can seen some of those inconsistencies:

https://streamable.com/l6ls2

To list a few ones: size of controls bigger or smaller than normal, the color of the UI not matching system theme, controls not rendered properly, a poor quality design in general. So I had no option but to choose the "best looking" style from all worsts, customize a bit and go with it to meet a short time development release.  
The difference between the old style and the new one is that the new one can adapt to the native color scheme of the system, also the controls can adapt to different screen sizes. Bellow you can see how Webcamoid looks with different color schemes, from light to dark schemes, including also to weird and fancy ones:

https://streamable.com/jq7fe

A lot more changes will come soon, this is just the very beginning.  
Ideas and critics are welcome :smile:

**Comment by gordonel** [comment]
Looks great, but I can't enjoy it since I can't compile webcamoid on my mac properly. You know why :)

**Comment by hipersayanX** [comment]
This is how Webcamoid looks now.

![screenshot](https://user-images.githubusercontent.com/1249885/75044347-3d68cb80-54a0-11ea-8117-fde7de0b9f63.png)

**Comment by hipersayanX** [comment]
Ok, with the style more or less defined is time to make the big real change. Let start with the main screen.

# Main screen - Photo capture

![00 - main 00](https://user-images.githubusercontent.com/1249885/75044543-92a4dd00-54a0-11ea-85da-65951b189e6f.png)

This is the very first screen you'll see when opening Webcamoid. Webcamoid will be ready to take a picture.  
In in the top left corner there will be an icon to apply video effects to the camera, desktop, video, etc.. Bellow it, an icon for enabling/disabling the camera flash (either software or hardware), and the option to set the timeout before taking the photo.  
In the top right corner is the settings icon that offers even more options, I'll talk about it later.  
In the bottom, you have access to the most important options for the current view.  
In the middle you've a big button to take a picture, in the left you can click/touch and have direct access to the last picture taken, in the right you have the recording icon that you can use to switch to the recording mode.  
And in the bottom you have a progress bar indicating the photo timeout.

# Main screen - Video recording

![00 - main 01](https://user-images.githubusercontent.com/1249885/75044564-9f293580-54a0-11ea-9430-65af47ae54e8.png)

When you touch the recording icon, the last photo preview will disappear, the take photo icon will be displaced to the left, the record icon to the middle, and the last video preview will appear in the right. The photo related controls disappears.  
When recording, a floating control will appear showing the current recording time.  

This design may works well in both, small and big screens.

**Comment by hipersayanX** [comment]
# Video effects

![01 - effects 00](https://user-images.githubusercontent.com/1249885/75044698-df88b380-54a0-11ea-9ba1-344b9a6fcdc4.png)

When you touch the effect icon, a drawer in the left side will appear showing all applied effects (clean by default). If you want to add a new effect you touch the homonym option. Each effect is listed as a swipe control, this is a control type that when you drag it you can discover new options, in this case options for configuring and removing the effect.

# Video effect selection

![01 - effects 01](https://user-images.githubusercontent.com/1249885/75044715-e7485800-54a0-11ea-8705-98c3ac0b418d.png)

When you press "Add effect", a dialog showing all available effects and a preview will appear. You can use the top left TextField to search the specific effect you want to use. You will also have the option to overlap the effect (apply a new effect over an applied one), if you don't want to overlap the effects, the new selected effect will replace the current one (default action). The overlap option is the equivalent of the old "Advanced effect mode" but more intuitive and easier to use.

# Video effect options

![01 - effects 02](https://user-images.githubusercontent.com/1249885/75044743-edd6cf80-54a0-11ea-84c5-68a868548e5a.png)

When you press "Configure", you will able to configure the effect, just like before.

**Comment by hipersayanX** [comment]
# Settings menu

![02 - settings menu](https://user-images.githubusercontent.com/1249885/75044916-2c6c8a00-54a1-11ea-8892-a4b870118c23.png)

The settings menu has the most important options ready to use. You can also play and stop the playback from there.

**Comment by hipersayanX** [comment]
# Sources bar

![03 - video settings 00](https://user-images.githubusercontent.com/1249885/75045040-56be4780-54a1-11ea-9be2-1e7982de9450.png)

If you select the video options, this drawer will appears, where you can use to set and configure the video sources and outputs, I'll talk about video outputs later.  
It will list all available webcams and the desktop by default, and you can add new sources like a video file or a GIF if you want.  

# Adding files

![03 - video settings 02](https://user-images.githubusercontent.com/1249885/75045057-5faf1900-54a1-11ea-878d-d2b79dce57bc.png)

You can add a file just like before.

# Adding multimedia streams

![03 - video settings 03](https://user-images.githubusercontent.com/1249885/75045080-68075400-54a1-11ea-9c81-836a006df7b6.png)

In the old _add file dialog_, it was already possible to add an URL but it was counter intuitive for many users, so this time _file and URL add_ is split between two tabs.

# Sources settings

![03 - video settings 04](https://user-images.githubusercontent.com/1249885/75045097-6fc6f880-54a1-11ea-9195-3506ad28a6c3.png)

And this is the setting menu for the source selected, similar to the old one.

**Comment by hipersayanX** [comment]
# Video outputs

![03 - video settings 01](https://user-images.githubusercontent.com/1249885/75045280-c0d6ec80-54a1-11ea-8571-ddd3676d5f9f.png)

Video outputs is the equivalent of the virtual camera but this time you have direct access to it. This will be used for virtual cameras but in the future is intended to be used with other kind of outputs like for example video streaming.

# Adding a new output

![03 - video settings 05](https://user-images.githubusercontent.com/1249885/75045300-c9c7be00-54a1-11ea-9a62-2d1b154d0b48.png)

You will be able to define which formats will be offered by your virtual camera.

# Advanced options

![03 - video settings 06](https://user-images.githubusercontent.com/1249885/75045324-d2b88f80-54a1-11ea-89c2-c0a08b20546d.png)

And you will be also able to control how the stream will be delivered to the client.

# Format dialog

![03 - video settings 07](https://user-images.githubusercontent.com/1249885/75045334-d9df9d80-54a1-11ea-9011-1fe04a0cf74f.png)

This dialog will appear when you want to define a new format.

**Comment by hipersayanX** [comment]
# Audio inputs

![04 - audio settings 00](https://user-images.githubusercontent.com/1249885/75045456-18755800-54a2-11ea-80f3-edca8b6e5020.png)

Audio settings menu brings you here, which is not much different from the old one.

# Audio outputs

![04 - audio settings 01](https://user-images.githubusercontent.com/1249885/75045464-1e6b3900-54a2-11ea-907d-52e495c8e7a9.png)

Here is the audio outputs menu

# Audio settings

![04 - audio settings 02](https://user-images.githubusercontent.com/1249885/75045486-24f9b080-54a2-11ea-86cc-628b879d9ba9.png)

And here you have the audio settings.   

There are not big changes in this section.

**Comment by hipersayanX** [comment]
# Image Capture

![05 - preferences 00](https://user-images.githubusercontent.com/1249885/75045632-668a5b80-54a2-11ea-9038-802a6cf191cf.png)

It will be possible to configure pictures saving options.

# Video Recording

![05 - preferences 01-00](https://user-images.githubusercontent.com/1249885/75045647-6ee29680-54a2-11ea-9426-e482cd5d5ee5.png)

Video saving options too.

# Video Recording - Codec options

![05 - preferences 01-01](https://user-images.githubusercontent.com/1249885/75045659-799d2b80-54a2-11ea-8c06-cb2f983d7f1b.png)

All options where they must be.

# General Options

![05 - preferences 02](https://user-images.githubusercontent.com/1249885/75045668-815cd000-54a2-11ea-85fd-659f7b5abbf6.png)

Same as before.

# Plugins - Paths

![05 - preferences 03-00](https://user-images.githubusercontent.com/1249885/75045699-8c176500-54a2-11ea-8aa9-4a9bb1555804.png)

One tab for plugins paths

# Plugins - Plugins

![05 - preferences 03-01](https://user-images.githubusercontent.com/1249885/75045710-92a5dc80-54a2-11ea-9f3b-4019ecc6447b.png)

and other for managing them.

# Updates

![05 - preferences 04](https://user-images.githubusercontent.com/1249885/75045720-98032700-54a2-11ea-94b6-04357e47ec86.png)

Same as before.

# About

![05 - preferences 05](https://user-images.githubusercontent.com/1249885/75045731-9df90800-54a2-11ea-8a4b-af3110f20437.png)

Same as before.

# Contributors

![05 - preferences 06](https://user-images.githubusercontent.com/1249885/75045749-a5b8ac80-54a2-11ea-9f33-1f5c83055f29.png)

# License

![05 - preferences 07](https://user-images.githubusercontent.com/1249885/75045766-abae8d80-54a2-11ea-9e6d-33b822864b39.png)

Same as before.

# 3rd Party Licenses

![05 - preferences 08](https://user-images.githubusercontent.com/1249885/75045781-b23d0500-54a2-11ea-8285-9850fe923f22.png)

Same as before.

**Comment by hipersayanX** [comment]
And that is my current plan for the UI redesign.

**Comment by hipersayanX** [comment]
Current progress:

![screenshot 01](https://user-images.githubusercontent.com/1249885/75912026-47d77d80-5e2f-11ea-86e0-7239650dbad0.png)

Effect button in the top left corner is dummy.
Video recording button in the bottom right corner is dummy.
**Video**, **Audio** and **Preferences** options in the setting menu are dummy. 
**Play sources** replaces the **Play/Stop** option in the toolbar.
Photo shoot is fully working. 

![screenshot 02](https://user-images.githubusercontent.com/1249885/75912030-49a14100-5e2f-11ea-9d15-5d95ba016236.png)

**Comment by hipersayanX** [comment]
More updates on the **Preferences** menu:

![00 - GeneralOptions](https://user-images.githubusercontent.com/1249885/76166878-dd7d5080-6140-11ea-8b9d-2e995fe5223b.png)
![01 - PluginsPaths](https://user-images.githubusercontent.com/1249885/76166879-de15e700-6140-11ea-9173-3618cf7d60be.png)
![02 - Plugins](https://user-images.githubusercontent.com/1249885/76166880-deae7d80-6140-11ea-9821-2688df2a0247.png)
![03 - Updates](https://user-images.githubusercontent.com/1249885/76166881-df471400-6140-11ea-9c65-67012270a3cb.png)
![04 - About](https://user-images.githubusercontent.com/1249885/76166882-dfdfaa80-6140-11ea-860c-87a89d0d98c5.png)
![05 - Contributors](https://user-images.githubusercontent.com/1249885/76166883-dfdfaa80-6140-11ea-9ae3-09b80555537d.png)
![06 - License](https://user-images.githubusercontent.com/1249885/76166884-e0784100-6140-11ea-8c38-ec9156d91e82.png)
![07 - ThirdPartyLicenses](https://user-images.githubusercontent.com/1249885/76166885-e110d780-6140-11ea-919d-a94233a9a1b0.png)

**Image Capture** and **Video Recording** are dummys.

**Comment by hipersayanX** [comment]
Image capture and video recording menus finished.

![00 - Image Capture](https://user-images.githubusercontent.com/1249885/77501944-3551d200-6e38-11ea-8d6b-e380034838fe.png)
![01 - Video Recording](https://user-images.githubusercontent.com/1249885/77501947-3682ff00-6e38-11ea-82f0-d18fece9e5ac.png)
![02 - Video Format Options](https://user-images.githubusercontent.com/1249885/77501948-371b9580-6e38-11ea-9d3a-0e1ed461b18f.png)
![03 - Video Codec Options](https://user-images.githubusercontent.com/1249885/77502062-85c92f80-6e38-11ea-9fdf-b84a4be4d7cf.png)
![04 - Audio Codec Options](https://user-images.githubusercontent.com/1249885/77501951-37b42c00-6e38-11ea-9b09-dce5b94f597d.png)

**Comment by hipersayanX** [comment]
Effects menus ready

![00 - Effects Panel](https://user-images.githubusercontent.com/1249885/80166190-be813380-85b3-11ea-8132-de7a945290df.png)
![01 - Effects Options](https://user-images.githubusercontent.com/1249885/80166211-cb058c00-85b3-11ea-8e0d-2a94c6660de0.png)
![02 - Add Effect](https://user-images.githubusercontent.com/1249885/80166227-d48ef400-85b3-11ea-95f3-410c0ab80a86.png)

**Comment by hipersayanX** [comment]
Audio settings ready.

![00 - Audio Settings](https://user-images.githubusercontent.com/1249885/80752112-5b3a5880-8b01-11ea-93a9-7f88fa5f1f56.png)
![01 - Audio Output Device Settings](https://user-images.githubusercontent.com/1249885/80752114-5c6b8580-8b01-11ea-8aab-17b0ab4d6ac2.png)

**Comment by hipersayanX** [comment]
Video sources ready.

![Video Sources](https://user-images.githubusercontent.com/1249885/81098530-e0e24d80-8edf-11ea-85d6-cc0e4ead092b.png)
![Video Source Options](https://user-images.githubusercontent.com/1249885/81098535-e2137a80-8edf-11ea-848b-ccd27a408e6d.png)

**Comment by hipersayanX** [comment]
Virtual camera working, with some bugs that must be fixed yet.

![00 - Outputs](https://user-images.githubusercontent.com/1249885/82738174-386c2000-9d0c-11ea-9c96-b00dcf6f00c2.png)

Virtual camera formats are now customizable .

![01 - Add Virtual Camera](https://user-images.githubusercontent.com/1249885/82738175-399d4d00-9d0c-11ea-9745-88e4df8793ba.png)
![02 - Add Format](https://user-images.githubusercontent.com/1249885/82738176-3a35e380-9d0c-11ea-813a-f85d00ea90da.png)
![03 - Virtual Camera](https://user-images.githubusercontent.com/1249885/82738179-3ace7a00-9d0c-11ea-94a4-76e896cae8ff.png)

You can also control the behavior in the client side.

![04 - Virtual Camera Options](https://user-images.githubusercontent.com/1249885/82738180-3b671080-9d0c-11ea-9f25-a204dd14dfaa.png)

Also, have removed the tray icon for announcing new version updates and have replaced by a dialog.

![05 - Updates](https://user-images.githubusercontent.com/1249885/82738181-3bffa700-9d0c-11ea-8c6e-d6e3cda05632.png)

**Comment by Slion** [comment]
Can't find that place in the settings where you can define virtual cam output resolution. I must be getting old. Running 8.8.0 on Windows.

**Comment by hipersayanX** [comment]
@Slion these changes are for Webcamoid 9 (currently daily build), Webcamoid 8 has not such options.

**Comment by hipersayanX** [comment]
And finally this is the last feature I want to add to Webcamoid 9. Webcamoid will advice to the user if the virtual camera is not installed or is outdated, and offer to download and install it. Alternatively it will inform if the virtual camera is not supported in your operating system.

![00 - vcam warnings](https://user-images.githubusercontent.com/1249885/132376820-a3fce9a2-98b5-487f-bd1e-00e58a23af38.png)

These will be more or less the download dialogs:

![01 - vcam download](https://user-images.githubusercontent.com/1249885/132376824-a6bc5155-cb0b-4aa5-aa4b-823587968888.png)

Once I finish with this feature, I'll start reviewing all pending issues and fixing bugs, and make the final release.

**Comment by hipersayanX** [comment]
And that's all,  every single feature I wanted for Webcamoid 9 is done, now is time to polish the code and make the final release.

![Screenshot_20210910_165624](https://user-images.githubusercontent.com/1249885/132911269-69694112-e329-421d-89c2-8cac570ffd56.png)
![Screenshot_20210910_165648](https://user-images.githubusercontent.com/1249885/132911271-53f55443-382a-4042-901d-637a64b06e68.png)

**Comment by cjbarth** [comment]
> And that's all, every single feature I wanted for Webcamoid 9 is done, now is time to polish the code and make the final release.

I would love to help with the final testing, but the nighties aren't building.

https://github.com/webcamoid/webcamoid/tree/daily-build

**Comment by hipersayanX** [comment]
@cjbarth because you are taking the tag as reference, the daily-build tag is fixed to an irrelevant commit, but the assets are build directly from [master](https://github.com/webcamoid/webcamoid), and right now is building just fine.
Afaik, isn't possible to create a tag that always points to the latest git commit, so I've created a tag pointing to a random commit and leaving the last commit message and a reference to the commit where all assets come from, this is the best solution I've found so far.

https://github.com/webcamoid/webcamoid/releases/tag/daily-build

**Comment by cjbarth** [comment]
I see... thus the confusion. Would something like this work?

https://github.com/marketplace/actions/deploy-nightly
https://github.com/marketplace/actions/daily-version

**Comment by hipersayanX** [comment]
@cjbarth will check, but for me is ok as is right now.

**Comment by shawnk0723** [comment]
Hi, I'm currently testing latest daily build (webcamoid-installer-linux-daily-master-1-x86_64.run) on arch linux. It says one of the updated feature from new UI design is "direct access to the last picture taken..." I do see the last photo preview icon but it's not clickable. Any idea what might be the issue? thank you for the great app!! :)

**Comment by shawnk0723** [comment]
> @cjbarth will check, but for me is ok as is right now.

Hi, I'm currently testing latest daily build (webcamoid-installer-linux-daily-master-1-x86_64.run) on arch linux. It says one of the updated feature from new UI design is "direct access to the last picture taken..." I do see the last photo preview icon but it's not clickable. Any idea what might be the issue? thank you for the great app!! :)

**Comment by hipersayanX** [comment]
@shawnk0723 The preview is in the bottom left corner

![Screenshot_20211211_170044](https://user-images.githubusercontent.com/1249885/145690141-8e7987d3-f693-4118-94bd-716b24e93de2.png)

>I do see the last photo preview icon but it's not clickable. Any idea what might be the issue? 

I did checked and it's working fine for me, try downloading the daily build again, it should open the picture in your default image viewer.

**Comment by shawnk0723** [comment]
> @shawnk0723 The preview is in the bottom left corner
> 
> ![Screenshot_20211211_170044](https://user-images.githubusercontent.com/1249885/145690141-8e7987d3-f693-4118-94bd-716b24e93de2.png)
> 
> > I do see the last photo preview icon but it's not clickable. Any idea what might be the issue?
> 
> I did checked and it's working fine for me, try downloading the daily build again, it should open the picture in your default image viewer.

Thank you for the quick reply. I'll try to download again and let you know. :)

**Comment by shawnk0723** [comment]
> @shawnk0723 The preview is in the bottom left corner
> 
> ![Screenshot_20211211_170044](https://user-images.githubusercontent.com/1249885/145690141-8e7987d3-f693-4118-94bd-716b24e93de2.png)
> 
> > I do see the last photo preview icon but it's not clickable. Any idea what might be the issue?
> 
> I did checked and it's working fine for me, try downloading the daily build again, it should open the picture in your default image viewer.

I just reinstalled fresh downloaded daily build. It still doesn't work even after cleaning all previous cached/config files. I remember it used to work on my system. I think it stopped working after trying latest stable version (8.8.0) and went back to daily build. I also have cheese webcam app. it has similar feature and it opens the photo when preview icon is clicked without any issues. Not sure why it stopped working... I really like the UI of webcamoid but this is the biggest showstopper for me now. :(

**Comment by shawnk0723** [comment]
I wonder if it's just linux specific bug since it works fine on your machine.

**Comment by hipersayanX** [comment]
@shawnk0723 

>I think it stopped working after trying latest stable version (8.8.0) and went back to daily build.

That feature is not affected by any configuration, so switching versions won't make any change. That feature only depends on which image viewer you have configured by default on your system.

>I wonder if it's just linux specific bug since it works fine on your machine.

Again, I'm using Arch Linux and it's working fine for me, so it's unlikely that this is a GNU/Linux general bug. Which distribution are you using? also, which desktop environment? maybe Wayland or Xorg?

**Comment by shawnk0723** [comment]
> @shawnk0723
> 
> > I think it stopped working after trying latest stable version (8.8.0) and went back to daily build.
> 
> That feature is not affected by any configuration, so switching versions won't make any change. That feature only depends on which image viewer you have configured by default on your system.
> 
> > I wonder if it's just linux specific bug since it works fine on your machine.
> 
> Again, I'm using Arch Linux and it's working fine for me, so it's unlikely that this is a GNU/Linux general bug. Which distribution are you using? also, which desktop environment? maybe Wayland or Xorg?

That is super weird! FYI, I'm currently using garuda, kde, x11. All I know is it used to work just fine up until I switched back & forth from 8.8.0 to daily build. My default image viewer is  LXimage. I changed it to gthumb just to see if it might fix the issue but no luck with this either. I also tried to troubleshoot by executing webcamoid.sh from terminal. This is the error i got: 
**********************************************************************************
qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
/usr/bin/xdg-open: line 613: 162175 Aborted                 (core dumped) kde-open${KDE_SESSION_VERSION} "$1"
**********************************************************************************

Then I found a solution from someone who previously had similar issue on google. His was fixed by installing qtcreator but unfortunately, it didn't work on my machine. yeah so, I'm still stuck with this same issue. Any suggestion would be greatly appreciated. Thanks man!

**Comment by hipersayanX** [comment]
@shawnk0723

>I also tried to troubleshoot by executing webcamoid.sh from terminal. This is the error i got:

~But you did executed Webcamoid on first place and it worked, then how did you executed it on first place? did you downloaded the AppImage, the installer, or the portable version?~

**EDIT**

I've downloaded and tested the daily build, and I'm experiencing the same exact bug, it could be a packaging problem. @shawnk0723 can you please open a bug for that?

**Comment by shawnk0723** [comment]
> @shawnk0723
> 
> > I also tried to troubleshoot by executing webcamoid.sh from terminal. This is the error i got:
> 
> ~But you did executed Webcamoid on first place and it worked, then how did you executed it on first place? did you downloaded the AppImage, the installer, or the portable version?~
> 
> **EDIT**
> 
> I've downloaded and tested the daily build, and I'm experiencing the same exact bug, it could be a packaging problem. @shawnk0723 can you please open a bug for that?

Oh, so it's happening on your machine too? yeah, i don't think it's anything big. might be packaging issue like you said bc it was  working just fine. I just opened a bug for this. Please let me know once this bug is fixed. I'll test it out for you soon as it's fixed. 
Thanks a lot for the awesome app once again hipersayanX!

**Comment by hipersayanX** [comment]
@shawnk0723 yep, I didn't noticed because building from sources doesn't throw this error :neutral_face: , I think it's probably because is missing some Qt plugin.

**Comment by shawnk0723** [comment]
Thank you for the quick reply. I tried with source code but still no luck. it's weird that all the other image viewers work except for gthumb. I guess it's my luck. LOL I'll just use other viewer for now. Thanks man. :)

**Comment by hipersayanX** [comment]
New UI released with Webcamoid 9.0.0, closing.

**Reply by beedell-roke** [reply]
@hipersayanX, although I realise that continuation of this discussion is unnecessary because the interface has been improved, are you able to state what the specific problems of the universal syle are? I intend to create some software that utilises Qt, and do not want to maintain custom sylisation unnecessarily, so I would like to report or convince you to any problems that are caused by it.

"http://github.com/webcamoid/webcamoid/discussions/517#discussion-4050075" is what caused me to ask. I am thankful.

**Reply by hipersayanX** [reply]
One of the problems was with running Qml 2 based apps in KDE. I wanted to show an icon in the buttons, [the feature was added in QtQuick.Controls 2.3](https://doc.qt.io/qt-5/qml-qtquick-controls2-abstractbutton.html#icon-prop), but KDE only supports a very old version of Qml, so icons were not show when using the default plasma styles, and the solution was either writing a theme from scratch or using one of the Qt defaults one: Fusion, Material, or Universal, all of them are crap, Universal was the less crap of them, and I force it in all platforms in Webcamoid 8 to keep a visual consistency and to have the features I wanted, and had to force a color palette because the theme did not adapted to the desktop color scheme.

The other problem I found was when porting Webcamoid to Android, when testing Webcamoid in the emulator or in a real device, all elements looked out of scale, either too big or too small, because Qml does not adapt well to the dpi.

There are other "small" problems, like for example when setting certain properties the program crashed, not because the property or the object was defined by me, but it was a Qt defined one. When porting to Qt6, did you see the animation when clicking the photo or video button? that animation stopped working properly.

Later is the language design problems, Qml does not implements a proper MVC model, it just mix the structure, styling, and logic as a whole one, so fo example for writing a new style for your app you must re-implement every single widget with its new style and logic from scratch, in contrast to HTML and CSS in which you just define a new CSS style.

And almost forgot, their constant language changes. Imagine you want to support plugins in your app, plugins that can add new graphical elements to the main interface, plugins that can be written by other people outside the project but still can be integrated to it, if Qml keep changing the rules, plugins developers will have to rewrite their plugins for working in your app just because for the whim of Qt developers of changing the API.

That's all I can remember for now, and maybe missing something else.
Qml was created in 2009, and for me is still not ready for serious usage. If you want to create an app with a dynamic interface, my serious recommendation, just go for HTML and create a webapp.

If you are not too ambitious and just want a traditional computer app and don't care about supporting mobile, using Qt Widgets is perfectly fine.
